push-messaging-and-notifications/main.js

```diff
@@ -0,0 +1,175 @@
+        // and save it to send a push message at a later date
+        var subscriptionId = subscription.subscriptionId;
+        var endpoint = subscription.endpoint;
+        sendToServer(subscriptionId, endpoint);
+        
+        pushButton.textContent = 'Disable Push Messages';
+        pushButton.disabled = false;
+      })
+      .catch(function(e) {
+        if (Notification.permission === 'denied') {
+          // The user denied the notification permission which
+          // means we failed to subscribe and the user will need
+          // to manually change the notification permission to
+          // subscribe to push messages
+          console.warn('Permission for Notifications was denied');
+          pushButton.disabled = true;
+        } else {
+          // A problem occurred with the subscription, this can
+          // often be down to an issue or lack of the gcm_sender_id
+          // and / or gcm_user_visible_only
+          console.error('Unable to subscribe to push.', e);
+          pushButton.disabled = false;
+        }
+      });
+  });
+}
+
+// Once the service worker is registered set the initial state
+function initialiseState() {
+  // Are Notifications supported?
+  if (!('Notification' in window)) {
+    console.warn('Notifications aren\'t supported.');
+    return;
+  }
+
+  // Check the current Notification permission.
+  // If its denied, it's a permanent block until the
+  // user changes the permission
+  if (Notification.permission === 'denied') {
+    console.warn('The user has blocked notifications.');
+    return;
+  }
+
+  // Check if push messaging is supported
+  if (!('PushManager' in window)) {
+    console.warn('Push messaging isn\'t supported.');
+    return;
+  }
+
+  // We need the service worker registration to check for a subscription
+  navigator.serviceWorker.ready.then(function(serviceWorkerRegistration) {
+    // Do we already have a push message subscription?
+    serviceWorkerRegistration.pushManager.getSubscription()
+      .then(function(subscription) {
+        // Enable any UI which subscribes / unsubscribes from
+        // push messages.
+        var pushButton = document.querySelector('.js-push-button');
+        pushButton.disabled = false;
+
+        if (!subscription) {
+          // We aren’t subscribed to push, so set UI
+          // to allow the user to enable push
+          return;
+        }
+
+        // Set your UI to show they have subscribed for
+        // push messages
+        pushButton.textContent = 'Disable Push Messages';
+        isPushEnabled = true;
+      })
+      .catch(function(err) {
+        console.warn('Error during getSubscription()', err);
+      });
+  });
+}
+
+
+window.addEventListener('load', function() {
+  var pushButton = document.querySelector('.js-push-button');
+  pushButton.addEventListener('click', function() {
+    if (isPushEnabled) {
+      unsubscribe();
+    } else {
+      subscribeToPushManager();
```

Comment:
`subscribeToPushManager` is an odd name. How about just `subscribe`, to parallel the existing `unsubscribe` function? Alternatively, change `unsubscribe` to `unsubscribeFromPushManager` to make it clear that they're doing the opposite things.